### ISSUE TITLE
set: shouldn't support 'select ... into outfile ... union select ... ' and 'select ... union select SQL_BUFFER_RESULT ... ' statements

### ISSUE THREAD

**Issue by ChenPeng2013** [pingcap/tidb#19097]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
drop table if exists t1, t2;
CREATE TABLE t1 (a int not null, b char (10) not null);
insert into t1 values(1,'a'),(2,'b'),(3,'c'),(3,'c');
CREATE TABLE t2 (a int not null, b char (10) not null);
insert into t2 values (3,'c'),(4,'d'),(5,'f'),(6,'e');
select a,b from t1 into outfile 'skr' union select a,b from t2;
select * from t1 union select SQL_BUFFER_RESULT * from t2;
```

### 2. What did you expect to see? (Required)
in MySQL or MariaDB
```
mysql> select a,b from t1 into outfile 'skr' union select a,b from t2;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'union select a,b from t2' at line 1
mysql> select * from t1 union select SQL_BUFFER_RESULT * from t2;
ERROR 1234 (42000): Incorrect usage/placement of 'SQL_BUFFER_RESULT'
```
### 3. What did you see instead (Required)
```
mysql> select a,b from t1 into outfile 'skr' union select a,b from t2;
+---+------+
| a | b    |
+---+------+
| 3 | c    |
| 4 | d    |
| 5 | f    |
| 6 | e    |
| 2 | b    |
| 1 | a    |
+---+------+
6 rows in set (0.00 sec)

mysql> select * from t1 union select SQL_BUFFER_RESULT * from t2;
+---+------+
| a | b    |
+---+------+
| 1 | a    |
| 2 | b    |
| 4 | d    |
| 5 | f    |
| 6 | e    |
| 3 | c    |
+---+------+
6 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```
Release Version: v4.0.0-beta.2-921-gb75a30fda
Edition: Community
Git Commit Hash: b75a30fda8eef5962e2fde1c6254ff02f330fd21
Git Branch: master
UTC Build Time: 2020-08-10 02:44:14
GoVersion: go1.14.6
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

**Comment by nullnotnil** [comment]
I am not sure if `SQL_BUFFER_RESULT` does anything in TiDB? I think this is just a simple parser compatibility issue. Fixing it will help ensure applications can migrate back to MySQL if needed.

**Comment by lzmhhh123** [comment]
It's a compatibility issue about the parser. So I move the sig/exec to sig/ddl and change type/bug to type/compatibility.

**Comment by AilinKid** [comment]
`SQL_BUFFER_RESULT` is introduced in https://github.com/pingcap/parser/pull/304 for compatibility, and it has nothing to do with the TiDB side.

**Comment by wjhuang2016** [comment]
For the second SQL, we can document that `SQL_BUFFER_RESULT` is ignored.

**Comment by AilinKid** [comment]
Reason:
This is a difference between TiDB and MySQL. For lock-clause and into-clause, MySQL will parse them as the suffix of the `query_expression` which is an embedded loop of the select statement.

So SQL like `select a,b from t1 into outfile 'skr' union select a,b from t2 ` is invalid in MySQL, because after the into-clause is parsed, the union clause will be unrecognized, which leads to error.

However, it will be parsed successfully because every single select-statement can be recognized with an into-clause suffix. The reason why the output is shown on the result rather than be stored in the file is: building an internal select for union statement will ignore the into-option.

As a conclusion above, we only support the into-option for single select-statement, better doc it on our website. If we wanna fully be compatible with this, it will take some time.

**Comment by AilinKid** [comment]
PTAL @wjhuang2016 @bb7133

**Comment by wjhuang2016** [comment]
I think we should make TiDB be compatible with MySQL

**Comment by bb7133** [comment]
Sounds that is not 'difficult/easy'.

What's the estimated work time to change the parser?

**Comment by wjhuang2016** [comment]
@AilinKid Any progress?

**Comment by AilinKid** [comment]
> @AilinKid Any progress?

@tangenta is resorted to helping with this.

**Comment by wjhuang2016** [comment]
@tangenta Any progress?

**Comment by tangenta** [comment]
I tried to extract `lock-clause` and `into-clause` to an upper level, but the shift/reduce conflicts is hard to resolve. Can we throw the syntax error in the semantic action?

**Comment by wjhuang2016** [comment]
> I tried to extract `lock-clause` and `into-clause` to an upper level, but the shift/reduce conflicts is hard to resolve. Can we throw the syntax error in the semantic action?

You can ask amao or kenny for help.

**Comment by AilinKid** [comment]
> I tried to extract `lock-clause` and `into-clause` to an upper level, but the shift/reduce conflicts is hard to resolve. Can we throw the syntax error in the semantic action?

yes, it's quite hard to solve. I suspect there will be some other work after the parser is finished.

**Comment by morgo** [comment]
Changing it to minor since it's a minor compatibility issue, not a bug.
